Zend/zend_gc.c

```diff
@@ -1204,6 +1258,8 @@ static void gc_compact(void)
 	}
 }
 
+/* For all roots marked purple, traverse the graph, marking referred objects grey.
+ * See MarkRoots() in Bacon & Rajan. */
```

Comment:
Also, for each traversed node, this decrements the refcount of child nodes. The purpose of marking grey is to remember which nodes where already traversed during this step.

On a higher level this simulates the effect (on refcounts) of deleting roots. Nodes that end up with a refcount > 0, and nodes reachable from them, are proven to not be part of a garbage cycle, and to be live. `gc_scan_roots()` will then restore the refcount of live nodes and mark other ones as garbage.

Reply:
I clarified this.

Reply:
Done.